exec/cnex/lib/binary.c

```diff
@@ -0,0 +1,478 @@
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cell.h"
+#include "exec.h"
+#include "number.h"
+#include "stack.h"
+#include "nstring.h"
+
+
+static void range_check32(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_INT32) || number_is_greater(x, BID_MAX_INT32)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+static void range_checkU32(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_UINT32) || number_is_greater(x, BID_MAX_UINT32)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+static void range_check64(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_INT64) || number_is_greater(x, BID_MAX_INT64)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+static void range_checkU64(TExecutor *exec, Number x)
+{
+    if (number_is_less(x, BID_MIN_UINT64) || number_is_greater(x, BID_MAX_UINT64)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+    if (!number_is_integer(x)) {
+        exec_rtl_raiseException(exec, "ValueRangeException", number_to_string(x), BID_ZERO);
+        return;
+    }
+}
+
+void binary_and32(TExecutor *exec)
+{
+    Number y = top(exec->stack)->number; pop(exec->stack);
+    Number x = top(exec->stack)->number; pop(exec->stack);
+
+    range_check32(exec, x);
+    range_check32(exec, y);
+    push(exec->stack, cell_fromNumber(number_from_uint32(number_to_uint32(x) & number_to_uint32(y))));
+}
+
+void binary_extract32(TExecutor *exec)
+{
+    Number w = top(exec->stack)->number; pop(exec->stack);
+    Number n = top(exec->stack)->number; pop(exec->stack);
+    Number x = top(exec->stack)->number; pop(exec->stack);
+
+    range_check32(exec, x);
+
+    range_checkU32(exec, n);
+    range_checkU32(exec, w);
+    unsigned int b = number_to_uint32(n);
+    if (b >= 32) {
+        push(exec->stack, cell_fromNumber(BID_ZERO));
+        return;
+    }
+    unsigned int v = number_to_uint32(w);
+    if (b + v > 32) {
+        v = 32 - b;
+    }
+    if (v == 32) {
+        push(exec->stack, cell_fromNumber(x));
+        return;
+    }
+    push(exec->stack, cell_fromNumber(number_from_uint32((number_to_uint32(x) >> b) & (1 << v) - number_to_uint32(BID_ONE))));
```

Comment:
Don't use `number_to_uint32(BID_ONE)`. That's just doing unnecessary work. Use the literal constant `1`.

Reply:
Fixed.